### DISCUSSION TITLE
Unable to create get request to get data from DB with join and display in response model

### DISCUSSION THREAD

**Discussion by indrajeet-sengupta** [fastapi/fastapi#6900] [Questions]
Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:
```DB_Connection.py


import os
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

url = 'mysql://{0}:{1}@{2}:{3}/{4}'.format(os.getenv('username'), os.getenv('password'),
                                           os.getenv('endpoint'), os.getenv('port'), os.getenv('database'))

SQLALCHEMY_DATABASE_URL = url
engine = create_engine(SQLALCHEMY_DATABASE_URL)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


```
```model.py

from sqlalchemy import Boolean, Column, Integer, String, DateTime, func, ForeignKey, Float, Time, Date
from sqlalchemy.orm import relationship

from DB_Connection import Base

class Salon(Base):
    __tablename__ = "salon"

    id = Column(Integer, primary_key=True, index=True)
    email = Column(String(100), unique=True, index=True)
    name = Column(String(50))
    company_name = Column(String(100), nullable=False)
    is_active = Column(Boolean, default=True)
    contact = Column(String(20))
    default_currency = Column(String(10), nullable=False)
    provides_membership = Column(Boolean, default=False)
    logo = Column(String(100))
    verified = Column(Boolean, default=False)
    created = Column(DateTime, server_default=func.now(), nullable=False)
    modified = Column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False)
    salon_security = relationship("Salon_Security", back_populates='salon')

class Salon_Security(Base):
    __tablename__ = 'salon_security'

    id = Column(Integer, primary_key=True, index=True)
    salon_id = Column(Integer, ForeignKey('salon.id'), nullable=False)
    salon_branch_id = Column(Integer, ForeignKey('salon_branches.id'))
    artist_id = Column(Integer, ForeignKey('artist.id'))
    email = Column(String(200), nullable=False)
    client_id = Column(String(200), nullable=False)
    client_secret = Column(String(200), nullable=False)
    password = Column(String(500), nullable=False, unique=True)
    created = Column(DateTime, server_default=func.now(), nullable=False)
    modified = Column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False)
    salon = relationship("Salon", back_populates='salon_security')
```

```Schema.py

from pydantic import BaseModel

class Salon_Security_Response(BaseModel):
    client_id: Optional[str]
    client_secret: Optional[str]

    class Config:
        orm_mode = True


class SalonResponse(BaseModel):
    email: Optional[str]
    company_name: Optional[str]
    contact: Optional[str]
    default_currency: Optional[str]
    is_active: Optional[str]
    verified: Optional[bool]
    provides_membership: Optional[bool]
    salon_security: Salon_Security_Response

    class Config:
        orm_mode = True
```

```routes.py
from fastapi import APIRouter
from fastapi import Depends


router = APIRouter()

@router.post('/new', response_model= SalonResponse)
async def create_salon(db: Session = Depends(DB_Connection.get_db)):
    result = db.query(Salon.email,Salon.company_name,Salon.contact,Salon.default_currency,
                               Salon.created,Salon.modified,Salon.is_active,Salon.verified,
                               Salon_Security.client_secret, Salon_Security.client_id)\
                .join(Salon_Security, Salon.id == Salon_Security.salon_id)\
                .filter(Salon.email == 'testing@testin.com').first()
    
    return result
```

### Description

I am trying to get the result in 

```
{
  "email": "string",
  "company_name": "string",
  "contact": "string",
  "default_currency": "string",
  "is_active": "string",
  "verified": true,
  "provides_membership": true,
  "salon_security": {
    "client_id": "string",
    "client_secret": "string"
  }
}
```
but I get an error when I hit the API:

pydantic.error_wrappers.ValidationError: 1 validation error for SalonResponse
response -> salon_security
  field required (type=value_error.missing)

Not sure why this is happening. 

Any help here would be appreciated. Please let me know the issue.

**Comment by indrajeet-sengupta** [comment]
@tiangolo Could you help me out here?

**Comment by Mause** [comment]
That isn't a self contained example at all - I have no idea how to run it to reproduce your issue. What DB and ORM are you actually using here?

**Comment by indrajeet-sengupta** [comment]
I am using mysql database and sqlalchemy orm.

**Comment by Mause** [comment]
At a guess, I would say you have no "salon security" (whatever that is) for the salon you are requesting.

**Comment by indrajeet-sengupta** [comment]
i tried printing result.client_secret and that did work but I don't understand why it is not able to parse it in the response model. The query and the query output are correct if I print them. The only issue I have is to generate the response using the response model. Maybe I am doing something wrong.

**Comment by Mause** [comment]
Okay, but that isn't the shape you declared? In your schemas, you said client_secret is inside salon_security, not at the root.

**Comment by indrajeet-sengupta** [comment]
Yes i want to print it inside the salon_security dictionary.

{
  "email": "string",
  "company_name": "string",
  "contact": "string",
  "default_currency": "string",
  "is_active": "string",
  "verified": true,
  "provides_membership": true,
  "salon_security": {
    "client_id": "string",
    "client_secret": "string"
  }
}

This is how I want my output to be.

That is why I have salon_security in the root which is referring to the class Salon_Security_Response.

How should I make this kind of output?

**Comment by Mause** [comment]
Something like this might be closer to what you want then:

```py
result = db.query(Salon).filter(Salon.email == 'testing@testin.com').first()
```

(though that does depend on how exactly you've setup sqlalchemy to be used)

**Comment by indrajeet-sengupta** [comment]
Salon and Salon_Security are two different tables and I want to perform a join query and get the results in the format I specified.

The above query you gave would give me data from the Salon table but would not give me anything from Salon_Security. I need some columns from the other table as well. The query I wrote seems to work fine. I only cannot parse it into the response model.

**Comment by Mause** [comment]
Did you actually test the code I gave you? Sqlalchemy does joins automatically if you declare them on your table, as you have done

**Comment by indrajeet-sengupta** [comment]
Yes got null for client_secret and client_id.

**Comment by Mause** [comment]
That is valid according to your schema?

**Comment by indrajeet-sengupta** [comment]
No, I should have retrieved and got the client_id and client_secret values in the response and not null values since both the tables are having relationship defined so they should have joined themselves as you said but I don't think that worked that way.

**Comment by indrajeet-sengupta** [comment]
Any help?

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! 👏  🙇 

I assume you solved it, thanks for closing the issue @indrajeet-sengupta  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
